### PR TITLE
Fix cast date type return wrong result

### PR DESCRIPTION
**Pull request by HangyuanLiu**
For #3213 
We have multiple date type, and we also need to cast between different date types.
 If not cast, it will cause problems when binarypredicate

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/analysis/DateLiteral.java [hunk 1]
```diff
@@ -319,7 +319,7 @@ protected void toThrift(TExprNode msg) {
 
     @Override
     protected Expr uncheckedCastTo(Type targetType) throws AnalysisException {
-        if (targetType.isDateType()) {
+        if (targetType.equals(this.type)) {
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/analysis/DateLiteral.java]: Are you sure only change this point is enough？
Do we need to change this point:
FunctionCallExpr::analyzeImpl
```
        Type[] args = fn.getArgs();
        if (args.length > 0) {
            // Implicitly cast all the children to match the function if necessary
            for (int i = 0; i < argTypes.length; ++i) {
                // For varargs, we must compare with the last type in callArgs.argTypes.
                int ix = Math.min(args.length - 1, i);
                if (!argTypes[i].matchesType(args[ix]) && !(
                        argTypes[i].isDateType() && args[ix].isDateType())) {
                    uncheckedCastChild(args[ix], i);
                    //if (argTypes[i] != args[ix]) castChild(args[ix], i);
                }
            }
        }
```

File: fe/src/test/java/org/apache/doris/analysis/DateLiteralTest.java [hunk 2]
```diff
@@ -55,4 +55,23 @@ public void TwoDigitYear() {
         }
         Assert.assertFalse(hasException);
     }
+
+    @Test
+    public void uncheckedCastTo() {
```
Comment (kangkaisen) [inline fe/src/test/java/org/apache/doris/analysis/DateLiteralTest.java]: Would better also add a test to `QueryPlanTest`.

**Review by kangkaisen — DISMISSED** [review]
+1. LGTM

File: fe/src/main/java/org/apache/doris/analysis/DateLiteral.java (hunk 1 above)
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/DateLiteral.java]: Why not return a new DateLiteral when targetType and this.type are different but both dateType.

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/DateLiteral.java [hunk 3]
```diff
@@ -319,7 +319,17 @@ protected void toThrift(TExprNode msg) {
 
     @Override
     protected Expr uncheckedCastTo(Type targetType) throws AnalysisException {
-        if (targetType.equals(this.type)) {
+        if (targetType.isDateType()) {
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/analysis/DateLiteral.java:322]: If isDateType is true. The type could only be DATE and DATETIME.

So one `if  else` if enough?

File: fe/src/main/java/org/apache/doris/analysis/DateLiteral.java [hunk 4]
```diff
@@ -320,6 +320,13 @@ protected void toThrift(TExprNode msg) {
     @Override
     protected Expr uncheckedCastTo(Type targetType) throws AnalysisException {
         if (targetType.isDateType()) {
+            if (targetType.equals(Type.DATE)) {
+                this.castToDate();                            
+            } else if (targetType.equals(Type.DATETIME)) {
+                this.type = Type.DATETIME;
```
Comment (chaoyli) [inline fe/src/main/java/org/apache/doris/analysis/DateLiteral.java:326]: If Date not to convert DateTime, is OK?

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/DateLiteral.java (hunk 4 above)
Comment (HangyuanLiu) [reply fe/src/main/java/org/apache/doris/analysis/DateLiteral.java:326]: Yes, is ok。And explicit type conver(cast expr) also go to this line

**Review by chaoyli — APPROVED** [review]

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**